src/context/market_product_group.tsx

```diff
@@ -0,0 +1,170 @@
+import { WalletContextState, useWallet } from '@solana/wallet-adapter-react'
+import { AccountInfo, PublicKey } from '@solana/web3.js'
+import React, { ReactNode, useContext, useEffect, useState, FC, useMemo } from 'react'
+import { MPG_ID as MAINNET_MPG_ID, MPs as MAINNET_MPs } from '../pages/TradeV3/perps/perpsConstants'
+import { MPG_ID as DEVNET_MPG_ID, MPs as DEVNET_MPs } from '../pages/TradeV3/perps/perpsConstantsDevnet'
+import { MarketProductGroup } from '../pages/TradeV3/perps/dexterity/accounts'
+import { useConnectionConfig } from './settings'
+import { useCrypto } from './crypto'
+import { Perp, Product } from 'gfx-perp-sdk'
+import useActivityTracker from '@/hooks/useActivityTracker'
+import useSolSub, { SubType } from '@/hooks/useSolSub'
+
+const ONE_MINUTE = 1000 * 60
+interface IPerpsInfoMpg {
+  marketProductGroup: MarketProductGroup
+  marketProductGroupKey: PublicKey
+  activeProduct: IActiveProduct
+  mpgRawData: AccountInfo<Buffer>
+  perpInstanceSdk: Perp
+  perpProductInstanceSdk: Product
+}
+
+export interface IActiveProduct {
+  id: string
+  orderbook_id: string
+  bids: string
+  asks: string
+  event_queue: string
+  tick_size: number
+  decimals: number
+  pairName: string
+}
+
+const MarketProductGroupContext = React.createContext<IPerpsInfoMpg | null>(null)
+
+export const useMpgConfig = (): IPerpsInfoMpg => {
+  const context = useContext(MarketProductGroupContext)
+
+  if (!context) throw new Error('Missing Market Product Group Context')
+  return context
+}
+
+export const MarketProductGroupProvider: FC<{ children: ReactNode }> = ({ children }) => {
+  const [currentMPG, setMPG] = useState<PublicKey>(new PublicKey(DEVNET_MPG_ID))
+  const [marketProductGroup, setMarketProductGroup] = useState<MarketProductGroup | null>(null)
+  const [rawData, setRawData] = useState<AccountInfo<Buffer>>(null)
+  const [activeProduct, setActiveProduct] = useState<IActiveProduct>(MAINNET_MPs[0])
+  const { isDevnet } = useCrypto()
+  const [initTesting, setInitTesting] = useState<boolean>(false)
+  const [perpInstanceSdk, setPerpInstanceSdk] = useState<Perp>(null)
+  const [perpProductInstanceSdk, setPerpProductInstanceSdk] = useState<Product>(null)
+
+  const wallet: WalletContextState = useWallet()
+
+  const { perpsConnection: mainnetConnection } = useConnectionConfig()
+  const MPG_ID = useMemo(() => (isDevnet ? DEVNET_MPG_ID : MAINNET_MPG_ID), [isDevnet])
+
+  const MPs = useMemo(() => (isDevnet ? DEVNET_MPs : MAINNET_MPs), [isDevnet])
+
+  const { on, off } = useSolSub()
+  const connection = useMemo(() => mainnetConnection, [])
+
+  useActivityTracker({
+    callbackOff: () => setIsCurrentTabActive(false),
+    callbackOn: () => setIsCurrentTabActive(true)
+  })
+  const [isCurrentTabActive, setIsCurrentTabActive] = useState(true)
+
+  const updateMPGDetails = async (mpgData, mpgRawData) => {
```

Comment:
U can put this function in useEffect to prevent reallocation

Reply:
This is used in two useEffects